crates/re_chunk/src/shuffle.rs

```diff
@@ -0,0 +1,379 @@
+    /// memory in order to make it contiguous.
+    /// Otherwise, only the offsets will be shuffled.
+    /// For most use cases, the small gains in cache locality are not worth the much higher cost
+    /// of copying the data around.
+    //
+    // TODO(cmc): `sort_if_unsorted(false)` followed by `sort_if_unsorted(true)` will not do
+    // anything, which is really not worth the hassle right now.
+    #[inline]
+    pub fn sort_if_unsorted(&mut self, make_contiguous: bool) {
+        if self.is_sorted() {
+            return;
+        }
+
+        re_tracing::profile_function!();
+
+        let now = std::time::Instant::now();
+
+        let swaps = {
+            re_tracing::profile_scope!("swaps");
+            let mut swaps = (0..self.row_ids.len()).collect::<Vec<_>>();
+            swaps.sort_by_key(|&i| self.row_ids[i]);
+            swaps
+        };
+
+        self.shuffle_with(&swaps, make_contiguous);
+
+        re_log::trace!(
+            entity_path = %self.entity_path,
+            num_rows = self.row_ids.len(),
+            elapsed = ?now.elapsed(),
+            "chunk sorted",
+        );
+
+        #[cfg(debug_assertions)]
+        #[allow(clippy::unwrap_used)] // dev only
+        self.sanity_check().unwrap();
+    }
+
+    /// Randomly shuffles the chunk using the given `seed`.
+    ///
+    /// If `make_contiguous` is `true`, the underlying arrow data will be copied and shuffled in
+    /// memory in order to make it contiguous.
+    /// Otherwise, only the offsets will be shuffled.
+    /// For most use cases, the small gains in cache locality are not worth the much higher cost
+    /// of copying the data around.
+    #[inline]
+    pub fn shuffle_random(&mut self, seed: u64, make_contiguous: bool) {
+        re_tracing::profile_function!();
+
+        let now = std::time::Instant::now();
+
+        use rand::{seq::SliceRandom as _, SeedableRng as _};
+        let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
+
+        let swaps = {
+            re_tracing::profile_scope!("swaps");
+            let mut swaps = (0..self.row_ids.len()).collect::<Vec<_>>();
+            swaps.shuffle(&mut rng);
+            swaps
+        };
+
+        self.shuffle_with(&swaps, make_contiguous);
+
+        re_log::trace!(
+            entity_path = %self.entity_path,
+            num_rows = self.row_ids.len(),
+            elapsed = ?now.elapsed(),
+            "chunk shuffled",
+        );
+    }
+
+    /// Shuffle the chunk according to the specified `swaps`.
+    ///
+    /// `swaps` is a slice that maps an implicit destination index to its explicit source index.
+    /// E.g. `swap[0] = 3` means that the entry at index `3` in the original chunk should be move to index `0`.
+    ///
+    /// If `make_contiguous` is `true`, the underlying arrow data will be copied and shuffled in
+    /// memory in order to make it contiguous.
+    /// Otherwise, only the offsets will be shuffled.
```

Comment:
> Otherwise, only the offsets will be shuffled.

I don't believe this is allowed for ListArray. Offsets must be monotonically increasing and dense -- the length of each array is `(offset[n+1] - offset[n])`

We could, however, do this with ListView instead.

Reply:
Oh yeah, nice catch. No idea why `arrow2` allows it :|

We're not going to get `ListView` into `arrow2` any time soon obviously, so I'll just remove the non-contiguous path and leave a TODO that links to [our `arrow-rs` migration ticket](https://github.com/rerun-io/rerun/issues/3741).